README.md
# Readme

## Description

A block-based content editor, with support for static html generation, in phoenix live view.

> #### Disclaimer {: .warning}
>
> **This library is still heavily in development.**
>
> Things will break and at least initially, they
> will not maintain backwards compatibility. In fact, the final, release-worthy version may never
> see the light of day.
>
> Use it at your own risk!

Currently supports the following blocks:

- a generic content-editable wrapped which can output any of:
  `<p>`, `<h1>`, `<h2>`, `<h3>`, `<ul>`, `<pre>`, `<blockquote>`
- a very badly styled table
- a code block with synthax highlighting for elixir only

## Installation and Usage

Add it to your dependencies in `mix.exs`:

```elixir
deps: [
  # ...
  {:philtre, "~> 0.12.2"}
  # ...
]
```

Include the styles somewhere in your application, for example, from `app.js`:

```typescript
import 'philtre/dist/index.css';
```

Or from `app.css`:

```css
@import 'philtre/dist/index.css';
```

Import and add the necessary hooks to your live view application

```typescript
import * as philtreHooks from 'philtre/src/hooks';

const liveSocket = new LiveSocket('/live', Socket, {
  hooks: { ...philtreHooks, ...yourHooks },
});
```

Render the page component inside one of your live views

```elixir
def mount(%{}, _session, socket) do
  {:ok, assign(socket, %{editor: Philtre.Editor.new()})}
end

def render(assigns) do
  ~H"""
  <button phx-click="save">Save</button>
  <.live_component
    module={Philtre.UI.Page}
    id={@editor.id}
    editor={@editor}
  />
  """
end

def handle_event("save", %{}, socket) do
  json = Philtre.Editor.serialize(socket.assigns.json)
  # Save the json however you please
  # Load into editor using Philtre.Editor.normalize/1
  inspect(json)
  {:noreply, socket}
end

def handle_info({:update, %Philtre.Editor{} = editor}, socket) do
  {:noreply, assign(socket, :editor, editor)}
end
```

## Developing using Playground

Playground is a locally setup, minimal phoenix application which loads the editor files using local paths, so they are always kept up to date and are even being watched by esbuild.

THis means it allows for live-reload development of hte library.

To start it, run `mix playground`

Note that editor pages are saved as files under `playground\priv\documents` so you should probably periodically clean them.


CHANGELOG.md
# Changelog

# 0.12.2

- [CHORE] Update mix dependencies
  - credo 1.6.0 -> 1.6.7
  - dialyxir 1.1.0 -> 1.2.0
  - ex_doc 0.28.0 -> 0.29
  - floki 0.30 -> 0.33
  - jason 1.3.0 -> 1.4.0
  - phoenix_live_view 0.17.0 -> 0.18.3
  - phoenix 1.6.0 -> 1.6.15
- [CHORE] Add and apply heex formatter
- [CHORE] Fix warnings emitted by new phoenix_live_view
- [CHORE] Update npm dependencies
  - @typescript-eslint/eslint-plugin 5.21.0 -> 5.41.0
  - cypress 9.6.0 -> 9.7.0
  - cypress-real-events 1.7.1. -> 1.7.2
  - esbuild 0.14.38 -> 0.15.12
  - eslint-plugin-prettier 4.0.0 -> 4.2.1
  - phoenix 1.6.7 -> 1.6.15
  - prismjs 1.28.0 -> 1.29.0
  - typescript 4.6.3 -> 4.8.4

# 0.12.1

- [FIX] Publish JS
- [DOCS] Add publishing instructions

Previous version didn't incoude prebuilt js and types

# 0.12.0

- [REWRITE] More agnostic block system where the ContentEditable is a base
  block through which other blocks communicate
- [FEATURE] Basic implementation for a list - not release-ready yet
- [FEATURE] Include library version in every editor

## 0.11.1

- [FEATURE] Add Javascript support for code block
- [FIX] Remove block functionatliy not actually being saved unless there is a
  manual update to another block
- [FIX] Move and rename `mix convert` -> `mix philtre.convert`

## 0.11.0

- [FEATURE] New, more general block json format

### Notes

A new format a block serializes into has been introduced. It's description is
available in the docs, unter the **outline** section.

A task to convert from old format into new is also available. You can use it by
running

```
mix philtre.convert path_where_your_files_are
```

If you need to manually convert, it should also be quite straightforward, from
looking at the documentation.

The reason for the new format is to support more generalized block structures
in the future.

## 0.10.2

- [QA] Simplify Playground endpoint
- [QA] Reorganize e2e tests around scopes
- [REFACTOR] Introduce common LiveBlock component
- [REFACTOR] Introduce common StaticBlock component
- [FIX] Transform being applied to soon due to placeholder space character in new cells
- [QA] Add blank `.credo.exs`
- [REFACTOR] Make ContentEditable less hacky
- [QA] Document ContentEditable

## 0.10.1

- [FIX] Adding a new empty h1, h2, h3, or li would crash
- [FIX] Better padding on li block
- [FEATURE] Shift + enter in /code block starts new block below
- [FEATURE] Code block auto-focuses when added
- [FEATURE] Navigation between blocks using tab and shift + tab
- [QA] Remove pageModel from integration tests
- [DOCS] Proposal for extensible block interface

## 0.10.0

- [FEATURE] Basic code block only supporing elixir synthax highlighting
- [FIX] Using button to add block after a code or table block fails
- [QA] Improve docs
- [FIX] Styles for `/code` block were not getting included correctly

## 0.9.4 2022-05-30

- [FIX] Bug with typing after inserting new line at the end of block
- [CHORE] Loosen floki dependency & update
- [CHORE] Update phoenix_live_view dependency
- [TWEAK] Use `n` to split lines in block component, instead of `<br/>`
- [TWEAK] Clean up redundant cells during block operations

## 0.9.3 2022-05-25

- fix publish + installation story

## 0.9.2 2022-05-22

- separate out scss into individual files
- style and restructure table component
- publish to hex and npm
- add basic setup readme

## 0.9.0 2022-05-21

- basic table component

## 0.8.1 2022-05-12

- various bugfixes related to blocks splitting and joining
- phoenix dependency update
- failed attempt configuration for heex formatter
- basic styling for PRE block

## 0.8.0 2022-05-08

There have been many changes since the last update

- complete engine rewrite
- e2e test suite
- conversion to library
- support for scss
- various misc additions and tweaks

## 0.7.1 2022-01-25

- reduce blocks after backspacing or updating, so they contain a minimal amount of cells

## 0.7.0 2022-01-25

- blockquote
- further codebase simplication
- removal of page struct
- bugfixes to backspace operation

## 0.6.1 2022-01-19

- simplification and improvements to internal API
- fix for a selection bug

## 0.6.0 2022-01-16

- introduce esbuild via npm
- introduce eslint
- introduce prettier
- make topbar an npm dependency

## 0.5.2 2022-01-08

- update elixir to 1.13.1 + all dependencies
- clear up credo issues
- add CI action workflow
- documentation improvements

## 0.5.0 2022-01-08

- block selection feature
- copy paste feature
- cleanup
- known bug: typing too quickly changes cursor position

## 0.4.0 2021-11-21

- block + cell based editor, supporting following blocks
  - p
  - h1,h2,h3
  - ul
  - with block downgrades, splitting, etc.

## 0.2.0 2021-11-07

- credo + pre-commit hook for credo
- dialyzer + specs + fixes + pre-commit hook for dialyzer

## 0.2.0 2021-11-06

- basic markdown support

## 0.1.0

- Basic article crud and routes
- Slug generation and uniqueness


mix.exs
defmodule Philtre.MixProject do
  use Mix.Project

  @source_url "https://github.com/begedin/philtre"
  @version "0.12.2"

  def project do
    [
      app: :philtre,
      description: "A block-style editor for live view",
      version: @version,
      elixir: "~> 1.13",
      elixirc_paths: elixirc_paths(Mix.env()),
      compilers: Mix.compilers(),
      aliases: aliases(),
      deps: deps(),
      preferred_cli_env: [
        "test.watch": :test
      ],
      package: package(),
      docs: docs(),
      source_url: @source_url,
      dialyzer: [plt_add_apps: [:mix]]
    ]
  end

  defp package do
    [
      maintainers: ["Nikola Begedin"],
      licenses: ["MIT"],
      files: [
        "lib",
        "dist",
        "mix.exs",
        ".formatter.exs",
        "README.md",
        "LICENSE",
        "CHANGELOG.md"
      ],
      links: %{
        Changelog: "https://hexdocs.pm/philtre/changelog.html",
        GitHub: "https://github.com/begedin/philtre"
      }
    ]
  end

  def application do
    []
  end

  defp elixirc_paths(:test), do: ["lib", "playground", "test/support"]
  defp elixirc_paths(:dev), do: ["lib", "playground"]
  defp elixirc_paths(_), do: ["lib"]

  defp deps do
    [
      {:credo, "~> 1.6.7", only: [:dev, :test], runtime: false, optional: true},
      {:dialyxir, "~> 1.2.0", only: [:dev, :test], runtime: false},
      {:ex_doc, "~> 0.29", only: :dev, runtime: false},
      {:floki, "~> 0.33"},
      {:jason, "~> 1.4.0"},
      {:mix_test_watch, "~> 1.0", only: [:test]},
      {:phoenix_html, "~> 3.0"},
      {:phoenix_live_reload, "~> 1.2", only: [:dev]},
      {:phoenix_live_view, "~> 0.18.3"},
      {:phoenix, "~> 1.6.15"},
      {:plug_cowboy, "~> 2.5.0"},
      {:uuid, "~> 1.1.0"}
    ]
  end

  def docs do
    [
      main: "readme",
      source_ref: @version,
      source_url: @source_url,
      groups_for_modules: [
        General: [
          Philtre.Editor,
          Philtre.Editor.Engine,
          Philtre.Editor.Serializer,
          Philtre.Editor.Utils,
          Philtre.UI.Page
        ],
        Blocks: [
          Philtre.Block.Code,
          Philtre.Block.Table,
          Philtre.Block.ContentEditable
        ],
        ContentEditable: [
          Philtre.Block.ContentEditable.Cell,
          Philtre.Block.ContentEditable.CleanEmptyCells,
          Philtre.Block.ContentEditable.Reduce,
          Philtre.Block.ContentEditable.Selection
        ],
        Playground: [
          Playground.App,
          Playground.Controller,
          Playground.Documents,
          Playground.Endpoint,
          Playground.Live.Edit,
          Playground.Live.Index,
          Playground.Live.New,
          Playground.Router,
          Playground.Router.Helpers,
          Playground.View
        ]
      ],
      extras: [
        "README.md",
        "CHANGELOG.md",
        "LICENSE",
        "additional_docs/outline_json_format.md": [title: "Block JSON Format"],
        "additional_docs/proposal_extensible_components.md": [
          title: "Extensible Components"
        ]
      ],
      groups_for_extras: [
        Outlines: [
          "additional_docs/outline_json_format.md"
        ],
        Proposals: [
          "additional_docs/proposal_extensible_components.md"
        ]
      ]
    ]
  end

  defp aliases do
    [
      playground: "run --no-halt -e 'Playground.App.run()'"
    ]
  end
end
